packages/lit-html/src/test/lit-html_test.ts

```diff
@@ -2185,6 +2186,22 @@ suite('lit-html', () => {
       }
     );
 
+    const syncAsyncDirective = directive(
+      class extends AsyncDirective {
+        render(x: string) {
+          this.setValue(x);
+          return noChange;
+        }
+      }
+    );
+
+    test('async directive can call setValue synchronously', () => {
+      assertRender(
+        html`<div foo=${syncAsyncDirective('test')}></div>`,
```

Comment:
Can you add the directive in ChildPart position also? I expect it's going to error in SSR.

Just looked at the code, and it's safe for attribute parts because we [no-op AttributePart._commitValue](https://github.com/lit/lit/blob/main/packages/lit-html/src/private-ssr-support.ts#L67) before running directives, but ChildPart doesn't have a convenient way to do that; we just call `resolveDirective` assuming it will just return a value and won't ever actually touch the dummy part, but then `setValue` in render actually does.

If that does error like I expect, then one option is to just say `setValue` can only be called synchronously from `update` not render, which would solve it. Alternatively, we could make a similar private-ssr-support helper to no-op `_$setValue` on ChildPart and return any committed value when resolving directives. This is probably the more robust bet.